From f0c158ffba4822f4278b3a96cdbd32205d445586 Mon Sep 17 00:00:00 2001
From: Evan Maddock <maddock.evan@vivaldi.net>
Date: Thu, 24 Jun 2021 23:26:20 -0400
Subject: [PATCH] Use libxml2 to get info from svg files for frame counts since
 librsvg doesn't let you access the description anymore.

Copied from https://github.com/tux4kids/t4kcommon/commit/99e9d3895b480d5998513592f6af25096c6d1c50, as it doesn't apply on this version.

Signed-off-by: Evan Maddock <maddock.evan@vivaldi.net>
---
 src/t4k_loaders.c | 43 ++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 42 insertions(+), 1 deletion(-)

diff --git a/src/t4k_loaders.c b/src/t4k_loaders.c
index 7248e35..695ec32 100644
--- a/src/t4k_loaders.c
+++ b/src/t4k_loaders.c
@@ -40,6 +40,8 @@ static void savePNG(SDL_Surface* surf,char* fn); //TODO this could be part of th
 #ifdef HAVE_RSVG
 #include<librsvg/rsvg.h>
 #include<librsvg/rsvg-cairo.h>
+#include<libxml/parser.h>
+#include<libxml/tree.h>
 #endif
 
 #define SOUNDS_DIR "sounds"
@@ -48,6 +50,7 @@ static void savePNG(SDL_Surface* surf,char* fn); //TODO this could be part of th
 /* local functions */
 
 #ifdef HAVE_RSVG
+int             get_number_of_frames_from_svg(const char *file_name);
 SDL_Surface*    load_svg(const char* file_name, int width, int height, const char* layer_name);
 sprite*         load_svg_sprite(const char* file_name, int width, int height);
 SDL_Surface*    render_svg_from_handle(RsvgHandle* file_handle, int width, int height, const char* layer_name);
@@ -160,6 +163,44 @@ const char* find_file(const char* base_name)
 }
 #ifdef HAVE_RSVG
 
+int get_number_of_frames_from_svg(const char *file_name) {
+  xmlDocPtr svgFile;
+  xmlNodePtr svgNode = NULL, nodeIterator = NULL;
+  int number_of_frames = 0, found = 0;
+
+  svgFile = xmlReadFile(file_name, NULL, XML_PARSE_RECOVER | XML_PARSE_NOERROR | XML_PARSE_NOWARNING);
+
+  /* If it's null something's really wrong because we're trying to load a sprite that doesn't exist */
+  if(svgFile == NULL) {
+      DEBUGMSG(debug_loaders, "get_number_of_frames_from_svg: couldn't load svgFile: %s\n", file_name);
+      return 0;
+  }
+
+  svgNode = xmlDocGetRootElement(svgFile);
+
+  /* If it's null then something's really wrong because there should be a root in every svg file... */
+  if(svgNode == NULL) {
+      DEBUGMSG(debug_loaders, "get_number_of_frames_from_svg: couldn't load the root from the svgFile: %s", file_name);
+      xmlFreeDoc(svgFile); /* be clean */
+      return 0;
+  }
+
+  nodeIterator = svgNode->children;
+  while(nodeIterator) {
+      if(xmlStrcasecmp(nodeIterator->name, (const xmlChar*)"desc") == 0) {
+          sscanf((const char*)xmlNodeGetContent(nodeIterator), "%d", &number_of_frames);
+          xmlFreeDoc(svgFile);
+          return number_of_frames;
+      }
+      nodeIterator = nodeIterator->next;
+  }
+
+  /* if we get here we had no description, which means something's really wrong */
+  DEBUGMSG(debug_loaders, "get_number_of_frames_from_svg: couldn't find the description frame number count from svgFile: %s", file_name);
+  xmlFreeDoc(svgFile);
+  return 0;
+}
+
 /* Load a layer of SVG file and resize it to given dimensions.
    If width or height is negative no resizing is applied.
    If layer = NULL then the whole image is loaded.
@@ -214,7 +255,7 @@ sprite* load_svg_sprite(const char* file_name, int width, int height)
   new_sprite->default_img = render_svg_from_handle(file_handle, width, height, "#default");
 
   /* get number of frames from description */
-  sscanf(rsvg_handle_get_desc(file_handle), "%d", &new_sprite->num_frames);
+  new_sprite->num_frames = get_number_of_frames_from_svg(file_name);
   DEBUGMSG(debug_loaders, "load_svg_sprite(): loading %d frames\n", new_sprite->num_frames);
 
   for(i = 0; i < new_sprite->num_frames; i++)
-- 
2.32.0

